HIP-6.md

```diff
@@ -0,0 +1,147 @@
+}
+```
+
+#### Storage types
+
+To be forward compatible with alternative storage modalities, the announcement protocol includes a "storage type" parameter.
+
+Upon adoption, the only supported storage type will be AWS S3, which has `storageType == 1`.
+
+For the S3 type, `_storageMetadata` should be the base64-encoded URL of the S3 bucket.
+
+Future storage types may be proposed in future HIPs. The content of `_storageMetadata` is expected to vary by type.
+
+#### ValidatorRegistry
+
+A contract that implements the `IValidatorRegistry` interface should be deployed to each chain.
+
+This contract stores validator registrations for that chain and mailbox contract.
+
+Validators for that domain and mailbox contract should register themselves so that they are discoverable by relayers.
+
+```
+interface IValidatorRegistry {
+    /// @notice Returns the local domain for validator registrations
+    function localDomain() external view returns (uint32);
+
+    /// @notice Returns the mailbox contract for validator registrations
+    function mailbox() external view returns (address);
+
+    /// @notice Returns a list of validators that have registered
+    function validators() external view returns (address[] memory);
+
+    /**
+     * @notice Returns a list of all registrations for all provided validators
+     * @param _validators The list of validators to get registrations for
+     * @return A list of validator addresses, and registered storage types and
+     * metadata. Note that the validators return value will only match
+     * _validators when each validator in _validators has exactly one
+     * registration.
+     */
+    function getValidatorRegistrations(
+        address[] calldata _validators,
+    ) external view returns (
+        address[] memory validators,
+        uint8[] memory storageTypes,
+        bytes[] memory storageMetadata
+    );
+
+   /**
+    * @notice Registers a validator
+    * @param _storageType An enum representing the modality used to store
+    * signed checkpoints (e.g. AWS S3)
+    * @param _storageMetadata Information encoding the location of signed
+    * checkpoints, specific to the storage modaility (e.g. S3 bucket URL)
+    * @param _signature The validator announcement attestation previously
+    * specified in this HIP
+    * @returns True upon success
+    */
+    function registerValidator(
+        uint8 _storageType,
+        bytes calldata _storageMetadata
+        bytes calldata _signature
+    ) external returns (bool);
+}
+```
+
+### **Rationale**
+
+#### Registration of multiple storage types
+
+This proposal allows validators to register multiple storage types.
+
+Validators may choose to post their signed checkpoints in multiple locations (e.g. S3, GCP, IPFS) in order to maximize liveness.
+
+#### Deregistration
+
+For simplicity, this proposal intentionally omits the ability to deregister a storage modality.
```

Comment:
Yes, though the previous entry will only be overwritten if they register with the same storage modaility.

Although actually I think the current proposal doesn't have a way to disambiguate which signature was more recent, so I guess it wouldn't be recommended.

If you did, someone could just re-submit your original announcement and overwrite your new location

Reply:
It would be nice to be able for Validators to revoke a previous registration, as well as support multiple registrations for the same storage modality (i.e. different regions of AWS)

Reply:
Care to make an argument for this as being part of the initial scope? IMO these features can and should be omitted so we can ship faster.

Reply:
It's worth noting that these features can easily be added in a backwards compatible way in the future.

Reply:
I think an example would be to specify multiple S3 buckets for availability reasons. Or if a validator wants to move from S3 to  GCS. I'd imagine supporting this would be trivial but I don't feel strongly

Reply:
Updated spec allows for multiple registrations, but still doesn't support de-registration.

Reply:
Is the plan then for the contract to be upgradable? If this contract is not upgradable then I feel decently strong that de-registration should be be supported, even if just at the contract level. It seems very trivial to add such a feature